sql/analyze.go

```diff
@@ -0,0 +1,1036 @@
+		case parser.GT:
+			// a < x OR a > y
+			if cmp == 0 {
+				// x = y
+				return &parser.ComparisonExpr{
+					Operator: parser.NE,
+					Left:     lcmp.Left,
+					Right:    lcmp.Right,
+				}, nil
+			} else if cmp == 1 {
+				// x > y
+				return parser.DBool(true), nil
+			}
+			return left, right
+		case parser.GE:
+			// a < x OR a >= y
+			if cmp == -1 {
+				// x < y
+				return left, right
+			}
+			// x >= y
+			return parser.DBool(true), nil
+		case parser.LT, parser.LE:
+			// a < x OR (a < y OR a <= y)
+			if cmp == 1 {
+				// x > y
+				return left, nil
+			}
+			// x < y
+			return right, nil
+		}
+
+	case parser.LE:
+		switch rcmp.Operator {
+		case parser.EQ:
+			// a <= x OR a = y
+			if cmp == -1 {
+				// x < y
+				return left, right
+			}
+			// x >= y
+			return left, nil
+		case parser.NE:
+			// a <= x OR a != y
+			if cmp != -1 {
+				// x >= y
+				return parser.DBool(true), nil
+			}
+			// x < y
+			return right, nil
+		case parser.GT, parser.GE:
+			// a <= x OR (a > y OR a >= y)
+			if cmp != -1 {
+				// x >= y
+				return parser.DBool(true), nil
+			}
+			// x < y
+			return left, right
+		case parser.LT, parser.LE:
+			// a <= x OR a < y
+			if cmp == -1 {
+				// x < y
+				return right, nil
+			}
+			// x >= y
+			return left, nil
+		}
+	}
+
+	return parser.DBool(true), nil
+}
+
+func simplifyOneOrInExpr(left, right *parser.ComparisonExpr) (parser.Expr, parser.Expr) {
+	switch left.Operator {
+	case parser.EQ:
+		switch right.Operator {
+		case parser.In:
+			left, right = right, left
+		default:
+			panic(fmt.Sprintf("unexpected operator: %s", right.Operator))
+		}
+		fallthrough
+
+	case parser.In:
+		tuple, ok := left.Right.(parser.DTuple)
+		if !ok {
+			return parser.DBool(true), nil
+		}
+
+		var tuple2 parser.DTuple
+		switch right.Operator {
+		case parser.EQ:
+			rdatum, rok := right.Right.(parser.Datum)
+			if !rok {
+				return parser.DBool(true), nil
+			}
+			tuple2 = parser.DTuple{rdatum}
+		case parser.In:
+			tuple2, ok = right.Right.(parser.DTuple)
+			if !ok {
+				return parser.DBool(true), nil
+			}
+		}
+
+		// We keep the tuples for an in expression in sorted order. So now we just
+		// merge the two sorted lists.
+		left.Right = mergeSorted(tuple, tuple2)
+		return left, nil
+	}
+
+	return left, right
+}
+
+func simplifyComparisonExpr(n *parser.ComparisonExpr) parser.Expr {
+	// NormalizeExpr will have left comparisons in the form "<var> <op>
+	// <datum>" unless they could not be simplified further in which case
+	// simplifyExpr cannot handle them. For example, "lower(a) = 'foo'"
+	if isVar(n.Left) && isDatum(n.Right) {
+		switch n.Operator {
+		case parser.EQ, parser.NE, parser.GT, parser.GE, parser.LT, parser.LE:
+			return n
+		case parser.In, parser.NotIn:
+			tuple, ok := n.Right.(parser.DTuple)
+			if !ok {
+				break
+			}
+			sort.Sort(tuple)
+			tuple = uniqTuple(tuple)
+			n.Right = tuple
+			return n
+		case parser.Like:
+			// a LIKE 'foo%' -> a >= "foo" AND a < "fop"
```

Comment:
No, `PrefixEnd("foo") == "fop"`. And that's the desired behavior. We want all keys with the prefix `"foo"`.